src/Foreign/Marshal/Alloc/Compat.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE ForeignFunctionInterface #-}
+module Foreign.Marshal.Alloc.Compat (
+  module Base
+, calloc
+, callocBytes
+) where
+import Foreign.Marshal.Alloc as Base
+
+#if !MIN_VERSION_base(4,8,0)
+import Foreign.C.Types
+import Foreign.Ptr (Ptr, nullPtr)
+import Foreign.Storable.Compat (Storable(..))
+import GHC.IO.Exception
+import Prelude.Compat
+
+-- |Like 'malloc' but memory is filled with bytes of value zero.
+--
+{-# INLINE calloc #-}
+calloc :: Storable a => IO (Ptr a)
+calloc = doCalloc undefined
```

Comment:
Some basic test for some of the functions in here would be nice, so that we at least have some evidence that the `foreign import` is correct.